File: internal/pkg/vormarun/manifest.go
package vormarun

import (
	"fmt"
	"io/fs"
	"strings"

	"github.com/vormadev/vorma/kit/bytesutil"
	"github.com/vormadev/vorma/kit/cryptoutil"
	"github.com/vormadev/vorma/kit/htmlutil"
	"github.com/vormadev/vorma/kit/jsonutil"
	"github.com/vormadev/vorma/kit/searchparams"
	"github.com/vormadev/vorma/kit/set"
)

// <dist>/.vorma/static/vorma.manifest.dev.json
const ManifestStaticOutDev = "vorma.manifest.dev.json"

// <dist>/.vorma/static/vorma.manifest.prod.json
const ManifestStaticOutProd = "vorma.manifest.prod.json"

type ClientModule struct {
	URL           string
	DepURLs       []string
	CSSBundleURLs []string
}

type ClientCoreAssets struct {
	ModuleURL string
	WasmURL   string
}

// Client Build ID is a hash of the JSON-marshaled Manifest
type Manifest struct {
	VormaVersion string

	// Dev-time
	Dev_ViteServerPort int    `json:",omitempty"`
	Dev_MuxPort        int    `json:",omitempty"`
	Dev_RefreshToken   string `json:",omitempty"`

	// Normalized config values needed at runtime
	PublicStaticBasePath string
	APIMountRoot         string
	UIVariant            string
	RootHTMLTemplateHash string

	// Static build outputs
	PublicFilepaths []string
	PublicFilemap   map[string]string
	CriticalCSS     string
	SearchSchemas   map[string]searchparams.Schema

	// Dev:    {
	//             URL: "http://localhost:5173/frontend/entry.tsx",
	//             DepURLs: [],
	//             CSSBundleURLs: [],
	//         }
	// Prod:   {
	//             URL: "/public/vorma_out_vite_frontend_entry-abc123.js",
	//             DepURLs: ["/public/vorma_out_vite_chunk_some_dep-abc123.js"],
	//             CSSBundleURLs: ["/public/vorma_out_vite_chunk_some_dep-abc123.css"],
	//         }
	ClientEntry      ClientModule
	ClientCoreAssets *ClientCoreAssets `json:",omitempty"`

	// Dev:    {
	//             "/user/:id": {
	//                 URL: "http://localhost:5173/frontend/user.tsx",
	//                 DepURLs: [],
	//                 CSSBundleURLs: [],
	//             },
	//         }
	// Prod:   {
	//             "/user/:id": {
	//                 URL: "/public/vorma_out_vite_frontend_user-abc123.js",
	//                 DepURLs: ["/public/vorma_out_vite_chunk_some_dep-abc123.js"],
	//                 CSSBundleURLs: ["/public/vorma_out_vite_chunk_some_dep-abc123.css"],
	//             },
	//         }
	ClientRoutes map[string]ClientModule
}

func (m Manifest) to_client_build_id() (string, error) {
	b, err := jsonutil.Serialize(m)
	if err != nil {
		return "", fmt.Errorf("error serializing manifest for hashing: %w", err)
	}
	hash := cryptoutil.Sha256Hash(b)
	b32 := bytesutil.ToBase32Raw(hash)
	return strings.ToLower(b32)[:24], nil
}

func (inst *Instance) PublicURL(src_path string) (string, error) {
	if IsBuild() {
		return "", nil
	}
	if err := inst.init(); err != nil {
		return "", fmt.Errorf("error initializing instance: %w", err)
	}
	pub_fm, err := inst.public_filemap()
	if err != nil {
		return "", fmt.Errorf("error getting public filemap: %w", err)
	}
	clean := strings.TrimPrefix(strings.TrimSpace(src_path), "/")
	url, ok := pub_fm[clean]
	if !ok {
		return "", fmt.Errorf("file %s not found in manifest public filemap", src_path)
	}
	return url, nil
}

func (inst *Instance) ClientBuildID() (string, error) {
	if IsBuild() {
		return "", nil
	}
	if err := inst.init(); err != nil {
		return "", fmt.Errorf("error initializing instance: %w", err)
	}
	if IsDev() {
		manifest, err := inst.manifest()
		if err != nil {
			return "", fmt.Errorf("error getting manifest: %w", err)
		}
		return manifest.to_client_build_id()
	}
	return inst.client_build_id_cache, nil
}

func (inst *Instance) CriticalCSSContentSha256() (string, error) {
	if IsBuild() {
		return "", nil
	}
	if err := inst.init(); err != nil {
		return "", fmt.Errorf("error initializing instance: %w", err)
	}
	el, err := inst.critical_css_el()
	if err != nil {
		return "", fmt.Errorf("error getting critical CSS element: %w", err)
	}
	hash, err := htmlutil.ComputeContentSha256(el)
	if err != nil {
		return "", fmt.Errorf("error hashing critical CSS element: %w", err)
	}
	return hash, nil
}

func (inst *Instance) PublicFS() (fs.FS, error) {
	if IsBuild() {
		return nil, nil
	}
	if err := inst.init(); err != nil {
		return nil, fmt.Errorf("error initializing instance: %w", err)
	}
	return fs.Sub(inst.static_fs, "public")
}

func (inst *Instance) public_static_base_path() (string, error) {
	manifest, err := inst.manifest()
	if err != nil {
		return "", fmt.Errorf("error getting manifest: %w", err)
	}
	return manifest.PublicStaticBasePath, nil
}

func (inst *Instance) critical_css_el() (*htmlutil.Element, error) {
	manifest, err := inst.manifest()
	if err != nil {
		return nil, fmt.Errorf("error getting manifest: %w", err)
	}
	return &htmlutil.Element{
		Tag:                 "style",
		AttributesKnownSafe: map[string]string{"id": Critical_CSS_EL_ID},
		DangerousInnerHTML:  manifest.CriticalCSS,
	}, nil
}

func (inst *Instance) public_filemap() (map[string]string, error) {
	manifest, err := inst.manifest()
	if err != nil {
		return nil, fmt.Errorf("error getting manifest: %w", err)
	}
	return manifest.PublicFilemap, nil
}

func (inst *Instance) final_public_filepaths() (*set.Set[string], error) {
	if IsDev() {
		return inst.make_final_public_filepaths()
	}
	return inst.final_public_filepaths_cache, nil
}

func (inst *Instance) make_final_public_filepaths() (*set.Set[string], error) {
	manifest, err := inst.manifest()
	if err != nil {
		return nil, fmt.Errorf("error getting manifest: %w", err)
	}
	filepaths := set.New[string]()
	for _, p := range manifest.PublicFilepaths {
		filepaths.Add(p)
	}
	return filepaths, nil
}

func (inst *Instance) manifest() (*Manifest, error) {
	if IsDev() {
		return read_manifest(inst.static_fs)
	}
	return inst.manifest_cache, nil
}

func read_manifest(static_fs fs.FS) (*Manifest, error) {
	out := ManifestStaticOutProd
	if IsDev() {
		out = ManifestStaticOutDev
	}
	manifest_bytes, err := fs.ReadFile(static_fs, out)
	if err != nil {
		return nil, fmt.Errorf("error reading manifest from static_fs: %w", err)
	}
	manifest, err := jsonutil.Parse[*Manifest](manifest_bytes)
	if err != nil {
		return nil, fmt.Errorf("error parsing manifest JSON: %w", err)
	}
	return manifest, nil
}

File: crates/vorma/src/manifest.rs
use std::collections::{BTreeMap, BTreeSet};

use data_encoding::BASE32_NOPAD;
use serde::{Deserialize, Serialize};

use crate::config::validate_public_static_base_against_api_mount;
use crate::constants::CRITICAL_CSS_EL_ID;
use crate::htmlutil::{Element, compute_content_sha256};

#[doc(hidden)]
pub const MANIFEST_STATIC_OUT_DEV: &str = "vorma.manifest.dev.json";
#[doc(hidden)]
pub const MANIFEST_STATIC_OUT_PROD: &str = "vorma.manifest.prod.json";

#[doc(hidden)]
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct ClientModule {
	pub url: String,
	pub dep_urls: Vec<String>,
	pub css_bundle_urls: Vec<String>,
}

#[doc(hidden)]
#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct ClientCoreAssets {
	pub module_url: String,
	pub wasm_url: String,
}

#[doc(hidden)]
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
pub struct Manifest {
	pub vorma_version: String,

	#[serde(default, skip_serializing_if = "is_zero")]
	pub dev_vite_server_port: i32,
	#[serde(default, skip_serializing_if = "is_zero")]
	pub dev_mux_port: i32,
	#[serde(default, skip_serializing_if = "String::is_empty")]
	pub dev_refresh_token: String,

	pub public_static_base_path: String,
	pub api_mount_root: String,
	pub ui_variant: String,
	pub root_document_shell_hash: String,

	pub public_filepaths: Vec<String>,
	pub public_filemap: BTreeMap<String, String>,
	pub critical_css: String,
	pub search_schemas: BTreeMap<String, serde_json::Value>,

	pub client_entry: ClientModule,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub client_core_assets: Option<ClientCoreAssets>,
	pub client_views: BTreeMap<String, ClientModule>,
}

impl Manifest {
	#[doc(hidden)]
	pub fn to_client_build_id(&self) -> Result<String, serde_json::Error> {
		let data = serde_json::to_vec(self)?;
		let hash = blake3::hash(&data);
		let mut encoded = BASE32_NOPAD.encode(hash.as_bytes());
		encoded.make_ascii_lowercase();
		encoded.truncate(24);
		Ok(encoded)
	}

	#[doc(hidden)]
	pub(crate) fn critical_css_el(&self) -> Element {
		Element {
			tag: "style".to_owned(),
			attributes: BTreeMap::from([("id".to_owned(), CRITICAL_CSS_EL_ID.to_owned())]),
			dangerous_inner_html: self.critical_css.clone(),
			..Element::default()
		}
	}

	#[doc(hidden)]
	pub fn critical_css_content_sha256(&self) -> String {
		compute_content_sha256(&self.critical_css_el())
	}

	#[doc(hidden)]
	pub fn public_url(&self, src_path: &str) -> Option<&str> {
		self.public_filemap
			.get(public_src_key(src_path)?)
			.map(String::as_str)
	}

	#[doc(hidden)]
	pub fn final_public_filepaths(&self) -> BTreeSet<String> {
		self.public_filepaths.iter().cloned().collect()
	}

	#[doc(hidden)]
	pub(crate) fn normalize_runtime_paths(&mut self) -> Result<(), String> {
		let (public_static_base_path, api_mount_root) =
			validate_public_static_base_against_api_mount(
				&self.public_static_base_path,
				&self.api_mount_root,
			)?;
		self.public_static_base_path = public_static_base_path;
		self.api_mount_root = api_mount_root;
		Ok(())
	}
}

pub(crate) fn public_src_key(src_path: &str) -> Option<&str> {
	let clean = src_path.trim().trim_start_matches('/');
	if clean.is_empty() {
		return None;
	}
	Some(clean)
}

fn is_zero(value: &i32) -> bool {
	*value == 0
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::htmlutil::render_element;

	#[test]
	fn manifest_json_keys_are_snake_case() {
		let manifest = Manifest {
			vorma_version: "0.1.0".to_owned(),
			public_static_base_path: "/public/".to_owned(),
			api_mount_root: "/api/".to_owned(),
			ui_variant: "react".to_owned(),
			root_document_shell_hash: "shell-hash".to_owned(),
			public_filepaths: vec!["/public/app.css".to_owned()],
			public_filemap: BTreeMap::from([(
				"app.css".to_owned(),
				"/public/app_hash.css".to_owned(),
			)]),
			critical_css: "body{}".to_owned(),
			search_schemas: BTreeMap::new(),
			client_entry: ClientModule {
				url: "/public/entry.js".to_owned(),
				dep_urls: Vec::new(),
				css_bundle_urls: Vec::new(),
			},
			client_views: BTreeMap::new(),
			..Manifest::default()
		};

		let json = serde_json::to_value(&manifest).unwrap();

		assert!(json.get("vorma_version").is_some());
		assert!(json.get("public_static_base_path").is_some());
		assert!(json.get("api_mount_root").is_some());
		assert!(json.get("root_document_shell_hash").is_some());
		assert!(json.get("root_html_template_hash").is_none());
		assert!(json.get("client_entry").unwrap().get("url").is_some());
		assert!(json.get("client_entry").unwrap().get("dep_urls").is_some());
		assert!(
			json.get("client_entry")
				.unwrap()
				.get("css_bundle_urls")
				.is_some()
		);
	}

	#[test]
	fn client_build_id_is_stable_base32_blake3_of_json_manifest() {
		let manifest = Manifest {
			vorma_version: "0.1.0".to_owned(),
			root_document_shell_hash: "shell-hash".to_owned(),
			client_entry: ClientModule {
				url: "/entry.js".to_owned(),
				..ClientModule::default()
			},
			..Manifest::default()
		};

		let first = manifest.to_client_build_id().unwrap();
		let second = manifest.to_client_build_id().unwrap();

		assert_eq!(first, second);
		assert_eq!(first.len(), 24);
		assert!(
			first
				.chars()
				.all(|ch| ch.is_ascii_lowercase() || ch.is_ascii_digit())
		);
	}

	#[test]
	fn public_url_trims_leading_slash_and_whitespace() {
		let manifest = Manifest {
			public_filemap: BTreeMap::from([(
				"favicon.ico".to_owned(),
				"/public/favicon_hash.ico".to_owned(),
			)]),
			..Manifest::default()
		};

		assert_eq!(
			manifest.public_url(" /favicon.ico "),
			Some("/public/favicon_hash.ico")
		);
		assert_eq!(manifest.public_url(" "), None);
	}

	#[test]
	fn manifest_runtime_paths_reject_overlapping_public_static_base() {
		let mut manifest = Manifest {
			public_static_base_path: "/api/assets/".to_owned(),
			api_mount_root: "/api/".to_owned(),
			..Manifest::default()
		};

		assert_eq!(
			manifest.normalize_runtime_paths().unwrap_err(),
			"public_static_base \"/api/assets/\" must not be under api_base \"/api/\""
		);
	}

	#[test]
	fn critical_css_el_escapes_raw_text_end_tag() {
		let manifest = Manifest {
			critical_css: "body{}/* </StYlE><script>alert(1)</script> */".to_owned(),
			..Manifest::default()
		};

		let html = render_element(&manifest.critical_css_el()).unwrap();

		assert!(!html.to_ascii_lowercase().contains("</style><script>"));
		assert!(html.contains(r"\3C /StYlE><script>alert(1)</script>"));
		assert!(html.ends_with("</style>"));
	}
}
